src/libPersistence/Retriever.cpp

```diff
@@ -167,6 +167,171 @@ void Retriever::RetrieveTxBlocks(bool& result, const bool& wakeupForUpgrade) {
+      VCBlockSharedPtr vcblock;
+
+      if (!BlockStorage::GetBlockStorage().GetVCBlock(
+              std::get<BlockLinkIndex::BLOCKHASH>(blocklink), vcblock)) {
+        LOG_GENERAL(WARNING,
+                    "Could not find vc with blockHash "
+                        << std::get<BlockLinkIndex::BLOCKHASH>(blocklink));
+        result = false;
+        return;
+      }
+      m_mediator.m_node->UpdateDSCommiteeCompositionAfterVC(*vcblock,dsComm);
+      
+    } else if (std::get<BlockLinkIndex::BLOCKTYPE>(blocklink) ==
+               BlockType::FB) {
+      FallbackBlockSharedPtr fallbackwshardingstruct;
+      if (!BlockStorage::GetBlockStorage().GetFallbackBlock(
+              std::get<BlockLinkIndex::BLOCKHASH>(blocklink),
+              fallbackwshardingstruct)) {
+        LOG_GENERAL(WARNING,
+                    "Could not find vc with blockHash "
+                        << std::get<BlockLinkIndex::BLOCKHASH>(blocklink));
+        result = false;
+        return;
+      }
+      uint32_t shard_id =
+          fallbackwshardingstruct->m_fallbackblock.GetHeader().GetShardId();
+      const PubKey& leaderPubKey =
+          fallbackwshardingstruct->m_fallbackblock.GetHeader()
+              .GetLeaderPubKey();
+      const Peer& leaderNetworkInfo =
+          fallbackwshardingstruct->m_fallbackblock.GetHeader()
+              .GetLeaderNetworkInfo();
+      const DequeOfShard& shards = fallbackwshardingstruct->m_shards;
+      m_mediator.m_node->UpdateDSCommitteeAfterFallback(
+          shard_id, leaderPubKey, leaderNetworkInfo, dsComm, shards);
+    }
+
+    m_mediator.m_blocklinkchain.AddBlockLink(
+        std::get<BlockLinkIndex::INDEX>(blocklink),
+        std::get<BlockLinkIndex::DSINDEX>(blocklink),
+        std::get<BlockLinkIndex::BLOCKTYPE>(blocklink),
+        std::get<BlockLinkIndex::BLOCKHASH>(blocklink));
+  }
+
+  if (!toDelete) {
+    result = true;
+    return;
+  }
+
+  for (; blocklinkItr != blocklinks.end(); blocklinkItr++) {
+    const auto& blocklink = *blocklinkItr;
+    if (std::get<BlockLinkIndex::BLOCKTYPE>(blocklink) == BlockType::DS) {
+      if (BlockStorage::GetBlockStorage().DeleteDSBlock(
+              std::get<BlockLinkIndex::DSINDEX>(blocklink))) {
+        BlockStorage::GetBlockStorage().PutMetadata(MetaType::DSINCOMPLETED,
+                                                    {'0'});
+      }
+    } else if (std::get<BlockLinkIndex::BLOCKTYPE>(blocklink) ==
+               BlockType::VC) {
+      if (!BlockStorage::GetBlockStorage().DeleteVCBlock(
+              std::get<BlockLinkIndex::BLOCKHASH>(blocklink))) {
+        LOG_GENERAL(WARNING, "Could not delete VC block");
+      }
+    } else if (std::get<BlockLinkIndex::BLOCKTYPE>(blocklink) ==
+               BlockType::FB) {
+      if (!BlockStorage::GetBlockStorage().DeleteFallbackBlock(
+              std::get<BlockLinkIndex::BLOCKHASH>(blocklink))) {
+        LOG_GENERAL(WARNING, "Could not deleteLoop  FB block");
+      }
+    }
+  }
```

Comment:
When toDelete is true, it means not only blockstorage should be removed, but also new block should not be added to m_blocklinkchain. We should consider toDelete before calling m_blocklinkchain.AddBlockLink().